### PR TITLE
feature(packages/ui): `Combobox` component

### PR DESCRIPTION
**Pull request by 0xTxbi**
# Description
This PR introduces a new `Combobox` component to our `ui` package, aimed at improving user experience and accessibility over traditional select elements. It includes a dynamic search functionality, making it easier for users to filter and locate options within large datasets. It also supports full keyboard navigation (`enter`, Arrow keys, `escape`, `tab`) for enhanced accessibility compliance. 

Additional features include customizable prepend/append elements, disabled states, and clear visual feedback with check icons, all integrated with our `FieldLayout` for consistent styling.

Overall, it simplifies dropdown selection.

## ScreenGrab
Screenshot 2024-10-07 at 10 12 22

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [x] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [x] I have performed a self-review of my own code
- [x] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by julien51** [comment]
ok ok, this definetely "looks" good and useful, but I want to make sure it "blends" in our existing style. wdyt?

**Comment by 0xTxbi** [comment]
> ok ok, this definetely "looks" good and useful, but I want to make sure it "blends" in our existing style. wdyt?

good question, and yes it does blend with our existing style and branding! do check this [demo](https://www.loom.com/share/aab8d0516ee94ffe9f3680ee39ce8a4e)

**Comment by julien51** [comment]
ok I think we are missing the "more" options here. It is important because it lets us show a smaller number of options by default but if the use r wants more, then more are shown. Do you think you could add that?

**Comment by 0xTxbi** [comment]
> ok I think we are missing the "more" options here. It is important because it lets us show a smaller number of options by default but if the use r wants more, then more are shown. Do you think you could add that?

good point! will add that!

**Review by julien51 — COMMENTED** [review]
Hum, I am not sure I see the `moreOptions` option.
The goal here is to show a specifc subset of items by default (not just the first n...)

File: packages/ui/lib/components/Form/Combobox.tsx [hunk 1]
```diff
@@ -22,7 +22,8 @@ interface Option {
 }
 
 interface ComboboxProps {
-  options: Option[]
+  initialOptions: Option[]
```
Comment (julien51) [inline packages/ui/lib/components/Form/Combobox.tsx:25]: Let's just keep `options` here for backward compatibility!

Comment (julien51) [inline packages/ui/lib/components/Form/Combobox.tsx:25]:
```suggestion
  options: Option[]
```

File: packages/ui/lib/components/Form/Combobox.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+// Helper function to get a random subset of items from an array
+const getRandomItems = (array: Option[], count: number): Option[] => {
+  const shuffled = [...array].sort(() => 0.5 - Math.random())
+  return shuffled.slice(0, count)
+}
+
+// Default story with a mix of initial and more options
+export const Default = {
+  args: {
+    initialOptions: getRandomItems(allOptions, 5),
```
Comment (julien51) [inline packages/ui/lib/components/Form/Combobox.stories.tsx:41]: shouldn't that be `options` ?

File: packages/ui/lib/components/Form/Combobox.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    onSelect: (selected) => {
+      console.log('selected', selected)
+    },
+  },
+} satisfies Story
+
+// Story with a pre-selected option
+export const WithInitialSelected = {
+  args: {
+    initialOptions: getRandomItems(allOptions, 5),
```
Comment (julien51) [inline packages/ui/lib/components/Form/Combobox.stories.tsx:53]: same?

File: packages/ui/lib/components/Form/Combobox.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      console.log('selected', selected)
+    },
+  },
+} satisfies Story
+
+// Story with a large number of options to test scrolling behavior
+export const WithLongList = {
+  args: {
+    // Generate 10 initial options
+    initialOptions: getRandomItems(
```
Comment (julien51) [inline packages/ui/lib/components/Form/Combobox.stories.tsx:66]: same?

File: packages/ui/lib/components/Form/Combobox.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    onSelect: (selected) => {
+      console.log('selected', selected)
+    },
+  },
+} satisfies Story
+
+// Story without any additional options in the "More options" section
+export const WithoutMoreOptions = {
+  args: {
+    initialOptions: getRandomItems(allOptions, allOptions.length),
```
Comment (julien51) [inline packages/ui/lib/components/Form/Combobox.stories.tsx:90]: ditto

File: packages/ui/lib/components/Form/Combobox.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    onSelect: (selected) => {
+      console.log('selected', selected)
+    },
+  },
+} satisfies Story
+
+// Story with duplicate options to test handling of non-unique values
+export const WithDuplicates = {
+  args: {
+    initialOptions: getRandomItems(allOptions, 5),
```
Comment (julien51) [inline packages/ui/lib/components/Form/Combobox.stories.tsx:100]: and there as well?

**Review by julien51 — CHANGES_REQUESTED** [review]
I think the stories need to be updated, no?

**Comment by 0xTxbi** [comment]
> I think the stories need to be updated, no?

oh definitely. the push failed for some reason!

**Review by julien51 — APPROVED** [review]

**Pull request merged**